warp_drive/training/lightning-trainer.py

```diff
@@ -0,0 +1,520 @@
+            self.clip_grad_norm[policy] = self._get_config(
+                ["policy", policy, "clip_grad_norm"]
+            )
+            if self.clip_grad_norm[policy]:
+                self.max_grad_norm[policy] = self._get_config(
+                    ["policy", policy, "max_grad_norm"]
+                )
+            normalize_advantage = self._get_config(
+                ["policy", policy, "normalize_advantage"]
+            )
+            normalize_return = self._get_config(["policy", policy, "normalize_return"])
+            gamma = policy_config["gamma"]
+
+            if algorithm == "PPO":
+                clip_param = self._get_config(["policy", policy, "clip_param"])
+
+            if algorithm == "A2C":
+                # Advantage Actor-Critic
+                self.trainers[policy] = A2C(
+                    discount_factor_gamma=gamma,
+                    normalize_advantage=normalize_advantage,
+                    normalize_return=normalize_return,
+                    vf_loss_coeff=vf_loss_coeff,
+                    entropy_coeff=entropy_coeff,
+                )
+                logging.info(f"Initializing the A2C trainer for policy {policy}")
+            elif algorithm == "PPO":
+                # Proximal Policy Optimization
+                self.trainers[policy] = PPO(
+                    discount_factor_gamma=gamma,
+                    clip_param=clip_param,
+                    normalize_advantage=normalize_advantage,
+                    normalize_return=normalize_return,
+                    vf_loss_coeff=vf_loss_coeff,
+                    entropy_coeff=entropy_coeff,
+                )
+                logging.info(f"Initializing the PPO trainer for policy {policy}")
+            else:
+                raise NotImplementedError
+
+        # Push all the data and tensor arrays to the GPU
+        # upon resetting environments for the very first time.
+        self.cuda_envs.reset_all_envs()
+
+        # Create and push data placeholders to the device
+        create_and_push_data_placeholders(
+            self.cuda_envs,
+            self.policy_tag_to_agent_id_map,
+            self.training_batch_size_per_env,
+            self.create_separate_placeholders_for_each_policy,
+        )
+
+        # Register action placeholders
+        if self.create_separate_placeholders_for_each_policy:
+            for policy in self.policies:
+                sample_agent_id = self.policy_tag_to_agent_id_map[policy][0]
+                action_space = self.cuda_envs.env.action_space[sample_agent_id]
+
+                self._register_actions(action_space, suffix=f"_{policy}")
+        else:
+            sample_policy = self.policies[0]
+            sample_agent_id = self.policy_tag_to_agent_id_map[sample_policy][0]
+            action_space = self.cuda_envs.env.action_space[sample_agent_id]
+
+            self._register_actions(action_space)
+
+        # Performance Stats
+        self.perf_stats = PerfStats()
```

Comment:
Time measurement would fit nicely in a callback.

Reply:
Thanks, will look into callbacks